diff -up gtk+-2.19.1/gdk-pixbuf/gdk-pixbuf-io.c.lib64 gtk+-2.19.1/gdk-pixbuf/gdk-pixbuf-io.c
--- gtk+-2.19.1/gdk-pixbuf/gdk-pixbuf-io.c.lib64	2009-11-30 13:10:35.000000000 -0500
+++ gtk+-2.19.1/gdk-pixbuf/gdk-pixbuf-io.c	2009-11-30 20:54:21.842859303 -0500
@@ -305,7 +305,17 @@ gdk_pixbuf_get_module_file (void)
   gchar *result = g_strdup (g_getenv ("GDK_PIXBUF_MODULE_FILE"));
 
   if (!result)
-	  result = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", "gdk-pixbuf.loaders", NULL);
+  {
+	  result = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", 
+				     HOST, "gdk-pixbuf.loaders", NULL);
+	  if (!g_file_test (result, G_FILE_TEST_EXISTS))
+	  {
+		  g_free (result);
+		  
+		  result = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", 
+					     "gdk-pixbuf.loaders", NULL);
+	  }
+  }
 
   return result;
 }
diff -up gtk+-2.19.1/gdk-pixbuf/Makefile.am.lib64 gtk+-2.19.1/gdk-pixbuf/Makefile.am
--- gtk+-2.19.1/gdk-pixbuf/Makefile.am.lib64	2009-11-26 01:35:49.000000000 -0500
+++ gtk+-2.19.1/gdk-pixbuf/Makefile.am	2009-11-30 20:54:21.842859303 -0500
@@ -521,6 +521,7 @@ INCLUDES = \
 	-I$(top_srcdir)/gdk-pixbuf 		\
 	-I$(top_builddir)/gdk-pixbuf 		\
 	-DGTK_SYSCONFDIR=\"$(sysconfdir)\"	\
+	-DHOST=\"$(host_triplet)\"              \
 	-DGTK_VERSION=\"$(GTK_VERSION)\"	\
 	-DGTK_BINARY_VERSION=\"$(GTK_BINARY_VERSION)\"	\
 	-DGTK_PREFIX=\"$(prefix)\"		\
diff -up gtk+-2.19.1/gdk-pixbuf/Makefile.in.lib64 gtk+-2.19.1/gdk-pixbuf/Makefile.in
--- gtk+-2.19.1/gdk-pixbuf/Makefile.in.lib64	2009-11-30 20:54:58.058857713 -0500
+++ gtk+-2.19.1/gdk-pixbuf/Makefile.in	2009-11-30 20:55:21.902591753 -0500
@@ -1344,6 +1344,7 @@ INCLUDES = \
 	-I$(top_srcdir)/gdk-pixbuf 		\
 	-I$(top_builddir)/gdk-pixbuf 		\
 	-DGTK_SYSCONFDIR=\"$(sysconfdir)\"	\
+	-DHOST=\"$(host_triplet)\"              \
 	-DGTK_VERSION=\"$(GTK_VERSION)\"	\
 	-DGTK_BINARY_VERSION=\"$(GTK_BINARY_VERSION)\"	\
 	-DGTK_PREFIX=\"$(prefix)\"		\
diff -up gtk+-2.19.1/gtk/gtkrc.c.lib64 gtk+-2.19.1/gtk/gtkrc.c
--- gtk+-2.19.1/gtk/gtkrc.c.lib64	2009-11-26 01:35:50.000000000 -0500
+++ gtk+-2.19.1/gtk/gtkrc.c	2009-11-30 20:54:21.844826383 -0500
@@ -450,7 +450,17 @@ gtk_rc_get_im_module_file (void)
       if (im_module_file)
 	result = g_strdup (im_module_file);
       else
-	result = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", "gtk.immodules", NULL);
+        {
+          result = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", GTK_HOST, 
+                                     "gtk.immodules", NULL);
+          if (!g_file_test (result, G_FILE_TEST_EXISTS))
+            {
+              g_free (result);
+             
+              result = g_build_filename (GTK_SYSCONFDIR, "gtk-2.0", 
+                                        "gtk.immodules", NULL);
+            }
+        }
     }
 
   return result;
